man/vpsadminctl.8
.\" Manpage for vpsadminctl.
.\" Contact jakub.skokan@vpsfree.cz to correct errors or typos.
.TH vpsadminctl 8 "16 Apr 2014" "1.18.2" "vpsadminctl man page"
.SH NAME
vpsadminctl \- remote control for vpsAdmind

.SH SYNOPSIS
\fBvpsadminctl <command>\fR [\fB-pvh\fR] [\fB-s\fR \fIsocket\fR] ...

.SH DESCRIPTION
\fBvpsadminctl\fR interacts with vpsAdmind. It can view its status, reload
configuration, control its lifetime and more.
.PP
\fBvpsadminctl\fR needs vpsAdmind to be running and with enabled remote control at all
times except for using \fB-h\fR or \fB-v\fR.
.PP
\fBvpsAdmind\fR must be started using system init script or manually,
\fBvpsadminctl\fR cannot start it.

.SH COMMON OPTIONS
.TP
\fB-p\fR, \fB--parsable\fR
Use in scripts, output can be easily parsed.

.TP
\fB-s\fR, \fB--socket\fR \fIsocket\fR
Connect to specified \fIsocket\fR instead of default
\fI/var/run/vpsadmind.sock\fR.

.TP
\fB-v\fR, \fB--version\fR
Print version and exit.

.TP
\fB-h\fR, \fB--help\fR
Show help and exit.

.SH COMMANDS
\fBget\fR <\fIsubcommand\fR> [\fB-Hl\fR]
.RS
Access vpsAdmind's resources and properties. When used with option \fB-p\fR,
output is formatted and printed in JSON.

\fBSubcommands\fR
.RS
.TP
\fBconfig\fR [\fIsome.key\fR]
Read and print config. If no argument is specified, print whole config.
.PP
.RS
Using argument \fIsome.key\fR you can select only specific part of config
to be printed. Keys are separated by '\fB.\fR'.
.RE

.TP
\fBqueue\fR
List transactions queued for execution. Transactions whose dependencies are not
met yet are not listed, as it is impossible to know when they will be executed.
.RE

.TP
\fB-H\fR, \fB--no-header\fR
Suppress header row

.TP
\fB-l\fR, \fB--limit\fR \fIn\fR
Limit numer of listed queued transactions to \fIn\fR. Defaults to 50.

.RE


\fBinstall\fR [\fB--create\fR \fB--addr\fR \fIaddr\fR \fB--location\fR
\fIid\fR|\fIlabel\fR ...] | [\fB--no-create\fR ...]
.RS
Add node to cluster or update node in cluster.
.PP
By default, \fBinstall\fR assumes \fB--create\fR, \fB--no-propagate\fR,
\fB--no-generate-configs\fR and \fB--no-ssh-key\fR.
.PP
\fBvpsadminctl install\fR does not modify vpsAdmind's configuration file,
node \fIid\fR must be set in config manually.
.PP
When used with \fB-p\fR, \fBvpsadminctl install\fR returns \fIid\fR of newly
created or updated node.
.PP
\fBvpsadminctl\fR blocks until \fBinstall\fR is finished. It does not wait for
all nodes to regenerate configs though, if you choose to propagate changes.
.TP
\fB--id\fR \fIid\fR
Node \fIid\fR must be a number greater than zero. If \fIid\fR already exists
in database, node's settings are updated.
.PP
.RS
\fIid\fR is automatically assigned by auto increment if omitted or invalid.
.RE

.TP
\fB--name \fIname\fR
Node \fIname\fR.

.TP
\fB--role\fR \fIrole\fR
Node \fIrole\fR might be one of \fBnode\fR, \fBstorage\fR or \fBmailer\fR.
Node roles may have additional specific options. Defaults to \fBnode\fR.

.TP
\fB--location\fR \fIid\fR|\fIlabel\fR
Location \fIid\fR or \fIlabel\fR to put node in. Mandatory in \fB--create\fR
mode.

.TP
\fB--addr\fR \fIaddr\fR
IPv4 address of this node. Mandatory in \fB--create\fR mode.

.TP
\fB--\fR[\fBno-\fR]\fBcreate\fR
Create (or not) new node in database.

.TP
\fB--\fR[\fBno-\fR]\fBpropagate\fR
Propagate (or not) changes to whole cluster. Currently, this includes only
regeneration of known hosts on all nodes.

.TP
\fB--\fR[\fBno-\fR]\fBgenerate-configs\fR
Generate (or not) OpenVZ sample configs for this node. This creates
\fIve-*.conf-sample\fR in \fI/etc/vz/conf\fR.
.PP
.RS
Generate configs only after node filesystem \fItype\fR is set, as each type
may handle configs differently. If you are installing new ZFS node, you have to
run \fBvpsadminctl install\fR twice. First to set filesystem \fItype\fR and
second to correctly generate configs.
.RE

.TP
\fB--\fR[\fBno-\fR]\fBssh-key\fR
Generate (or not) public and private SSH keys, if they do not exists. This
removes existing keys on this node and replaces them by the shared keys of the
cluster, so that every node can ssh every other node.

.TP
\fBOptions for role \fInode\fR
.RS
.TP
\fB--maxvps\fR \fIcount\fR
Maximum \fIcount\fR of VPS on this node. Defaults to \fB30\fR.

.TP
\fB--ve-private\fR \fIprivate\fR
Path to \fB$VE_PRIVATE\fR. \fB%{veid}\fR is always replaced by current VPS ID.
It is crucial that \fIprivate\fR is correctly set.

.TP
\fB--fstype\fR \fItype\fR
Filesystem type of this node. It might be one of \fBext4\fR, \fBzfs\fR or
\fBzfs_compat\fR. Defaults to \fBext4\fR.
.PP
If you use \fBzfs\fR or \fBzfs_compat\fR, \fIprivate\fR should be in
a subdirectory, for example \fI/vz/private/%{veid}/private\fR, because
\fBvzctl\fP(8) is not ready for it. For \fBext4\fR, use
\fI/vz/private/%{veid}\fR.
.RE

.RE


\fBkill\fR [[\fB-a\fR] | [\fB-t\fR]] [\fIid\fR|\fItype\fR]\fB...\fR
.RS
Kill selected running transactions. This command accepts a list of transaction
\fIids\fR or \fItypes\fR. Arguments are by default treated as transaction
\fIids\fR. Option \fB-t\fR changes that to transaction \fItype\fR.
.PP
This command does not kill transactions
waiting in queue, only those which are currently running.
.PP
Transactions are marked as failed, their error message set to "Killed".
.PP
\fBvpsadminctl\fR blocks until all matching transactions are killed.

.TP
\fB-a\fR, \fB--all\fR
Kill all running transactions, you do not have to provide list of \fIids\fR or
\fItypes\fR.

.TP
\fB-t\fR, \fB--type\fR
Arguments are transaction \fItypes\fR, not \fIids\fR.
.RE


\fBpause\fR [\fIid\fR]
.RS
Pause execution of queued transactions. Running transactions are finished,
new transactions are not executed until \fBvpsadminctl resume\fR is called.
.PP
Argument \fIid\fR is optional. If used, execution will be paused after
transaction with id \fIid\fR is finished. Until then, new transactions are
normally executed.
.PP
\fBpause\fR stops the execution of urgent transactions too.
.PP
\fBpause\fR returns immediately. It does not wait for the pause to take effect.
.RE


\fBrefresh\fR
.RS
Update info about this node, including kernel version, and all its VPSes and
datasets. Traffic accounting is not updated.
.PP
\fBvpsadminctl\fR blocks until refresh is finished.
.RE


\fBreinit\fR
.RS
Drops all iptable rules and chains concerning traffic accounting, then create
them again.
.PP
\fBvpsadminctl\fR blocks until reinit is finished.
.RE


\fBreload\fR
.RS
Instructs vpsAdmind to reload its configuration file.
.PP
\fBvpsadminctl\fR does not wait for the reload to actually finish, although it happens
instantly.
.RE


\fBrestart\fR [\fB-f\fR]
.RS
Order vpsAdmind to restart. vpsAdmind will wait for transactions that are running
to finish. It will not execute more transactions.
.PP
\fBvpsadminctl\fR does not wait for the restart to finish, it returns immediately.

.TP
\fB-f\fR, \fB--force\fR
Restart vpsAdmind immediately, do not wait for transactions to finish.
All transactions are softly killed and will restart when vpsAdmind is started.
.RE


\fBresume\fR
.RS
Resume transaction execution after it has been paused by \fBvpsadminctl pause\fR.
\fBresume\fR cancels both immediate and delayed \fBpause\fR.
.PP
\fBresume\fR can also be used to cancel scheduled \fBstop\fR, \fBrestart\fR or
\fBupdate\fR.
.PP
One useful side-effect of \fBresume\fR is that it forces \fBvpsAdmind\fR to check queue
for new transactions, no matter of database last change. In some rare cases,
\fBvpsAdmind\fR may seem to be stuck because it is ignoring queued transactions.
\fBresume\fR can be used to put \fBvpsAdmind\fR in motion again, until the
actual bug is resolved.
.PP
\fBresume\fR returns immediately.
.RE


\fBset\fR <\fIsubcommand\fR>
.RS
Change vpsAdmind's properties. \fBset\fR changes only runtime configuration,
nothing is written to disk.

\fBSubcommands\fR
.RS

.TP
\fBconfig\fR <\fIsome.key\fR\fB=\fR\fIvalue\fR>...
Alter vpsAdmind's configuration. Set key \fIsome.key\fR to \fIvalue\fR. Format of
keys is the same as for \fBget\fR. Multiple keys may be specified, separated
by spaces.

.RE
.RE


\fBstatus\fR [\fB-Hcw\fR]
.RS
Show vpsAdmind's status. If no option is specified, summary is shown.

.TP
\fB-H\fR, \fB--no-header\fR
Do not print header row, useful for scripts.

.TP
\fB-c\fR, \fB--consoles\fR
List exported consoles. Consoles are exported when
accessed from vpsAdmin web interface.

.TP
\fB-w\fR, \fB--workers\fR
List transactions and commands that are currently being run.
.RE


\fBstop\fR [\fB-f\fR]
.RS
Order vpsAdmind to exit. vpsAdmind will wait for transactions that are running
to finish. It will not execute more transactions.
.PP
\fBvpsadminctl\fR does not wait for the stop to finish, it returns immediately.

.TP
\fB-f\fR, \fB--force\fR
Stop vpsAdmind immediately, do not wait for transactions to finish.
All transactions are softly killed and will restart when vpsAdmind is started
later.
.RE


\fBupdate\fR [\fB-f\fR]
.RS
Stop vpsAdmind, execute git pull and start it again. This command behaves the
same as soft \fIstop\fR. It waits for running transactions to finish.
.PP
\fBvpsadminctl\fR does not wait for the update to finish, it returns immediately.

.TP
\fB-f\fR, \fB--force\fR
Update vpsAdmind immediately, do not wait for transactions to finish.
All transactions are softly killed and will restart when vpsAdmind is restarted.

.RE


.SH EXAMPLES
\fBStandard usage\fR
.PP
Check how vpsAdmind is doing:
  vpsadminctl status

Show what transactions and commands are running at the moment:
  vpsadminctl status -w

Kill two transactions you want to cancel. \fI1234\fR and \fI5678\fR are
transaction \fIids\fR, first column from \fBvpsadminctl status\fR:
  vpsadminctl kill 1234 5678

Kill all transactions:
  vpsadminctl kill -a

Kill all backups (\fI5005\fR is type of regular backup, \fI5006\fR of on-demand
backup, fourth column from \fBvpsadminctl status\fR):
  vpsadminctl kill -t 5005 5006

Update vpsAdmind. Do not forget to update also web interface and database scheme
afterwards:
  vpsadminctl update

Read server ID:
  vpsadminctl get config vpsadmin.server_id

Change number of concurrent workers:
  vpsadminctl set config vpsadmin.threads=10

Confirm change:
  vpsadminctl get config vpsadmin.threads

See what transactions are queued and will be executed, limit count to 10:
  vpsadminctl get queue -l 10

\fBInstall new ZFS node\fR
.PP
First, configure vpsAdmind so it can access database and start it. It will
complain that node is not registered, ignore that.

Create dataset vz/private and mount it to /vz/private.

When vpsAdmind is running, register node in database, do not create configs:

  vpsadminctl install --name node1 --role node --location 1 --addr 1.2.3.4
                      --maxvps 30 --fstype zfs_compat
                      --ve-private /vz/private/%{veid}/private
                      --no-propagate --no-generate-configs --no-ssh-key

Now, reload config to discover filesystem type as we have set it:

  vpsadminctl reload

Give it some time, as \fBvpsadminctl reload\fR does not block:

  sleep 5

Run install again, this time generate configs and keys:

  vpsadminctl install --no-create --propagate --ssh-key --generate-configs

Node should be installed and correctly configured.

\fBInstall new EXT4 node\fR
.PP
One \fBvpsadminctl install\fR will suffice for now, because ext4 is the default filesystem
\fItype\fR in config.

First steps are the same. Configure and start vpsAdmind, then run:

  vpsadminctl install --name node1 --role node --location 1 --addr 1.2.3.4
                      --maxvps 30 --fstype ext4
                      --ve-private /vz/private/%{veid}
                      --propagate --generate-configs --ssh-key

Node should be installed and correctly configured.

.SH AUTHOR
Jakub Skokan <jakub.skokan@vpsfree.cz>

.SH LICENSE
\fBvpsadminctl\fR is a part of vpsAdmin project and is released under GNU/GPLv2.


CHANGELOG
* Mon Apr 21 2014 - version 1.20.0
- No change

* Sun Apr 20 2014 - version 1.19.0
- Transaction add: check for duplicities
- VPS: fix redirect loop that sometimes happened on start/stop/restart
- Networking: sum total in and out in live monitor, fix history listing for admins
- Fix automated node selection algorithm to exclude node in maintenance and order better

* Tue Apr 15 2014 - version 1.18.2
- Status page: fix formatting when node count is odd
- Live network monitor: improve formatting, align to right, use full width
- Node maintenance toggle: move to extra column

* Fri Apr 11 2014 - version 1.18.0
- Urgent transactions

* Thu Apr 10 2014 - version 1.17.1
- Fix slow live network monitor and negative values

* Web Apr 09 2014 - version 1.17.0
- Track kernel version
- VPS migration: start VPS afterwards if it should be online or is online
  at the moment
- Live network monitor
- Per-node maintenance mode
- NAS export add: improve path validation
- Pages networking and adminvps: show index for admins, do not list

* Sun March 23 2014 - version 1.16.0
- Remove deprecated mail log
- General settings: add base URL
- Fix & update DB scheme
- Fix transactions box refreshing in Chrome
- Transactions: filter by done, success; redo transaction, user can filter
- Mass management: fix select all/none/reverse, filter by DNS resolvers, add backup
- Status page: rename event log, show only active members
- Config boolean to toggle online migrations
- NAS: add redirects after actions
- Member edit: add switch context link
- Login: prevent redirect loop for jumpto.php
- VPS swap: do not list lazy deleted VPS, fix backuper and config swapping
- VPS passwd: show generated password till logout
- VPS hostname: max length set to 255
- VPS clone to playground: fix backup export setting, fix expiration
- Transactions created in web UI have priority set to 10
- Backup restore: add missing check if user selected backup to restore from
- User requests: fix deny/ignore/invalidate for registration requests
- /cluster_status.php reports cluster state
- Transaction fallbacks
- Save only one server status, change storage to memory

* Fri Oct 4 2013 - version 1.15.1
- Fix node selection in approval request details
- Complete DB scheme
- Requests for approval: do not allow duplicit actions
- VPS start/stop/restart: redirect afterwards

* Fri Sep 27 2013 - version 1.15.0
- Jump box in header to quickly find members and VPS
- Regain admin privileges after dropping them without the need to relogin
- VPS passwd: show password also in the change form
- Force setting backup export when creating playground VPS
- Requests for approval system
- Templates: add flag supported and order
- Login: do not redirect back to login page

* Tue Sep 10 2013 - version 1.14.0
- Mass management: filter by config, new action change config
- VPS details: show used RAM and HDD
- Cluster config edit: leave checkbox 'reconfigure all affected VPS' unchecked by default
- Login: redirect user to page he came to
- ZFS backups, clone and migration: new ZFS compatibility mode

* Sat Sep 07 2013 - version 1.13.0
- Configuration form for daily reports

* Fri Sep 06 2013 - version 1.12.3
- Fix VPS online and offline migration return value

* Fri Jul 26 2013 - version 1.12.1
- ZFS restore: fix transaction addition when on-demand backup is not checked

* Thu Jul 18 2013 - version 1.12.0
- Online migration

* Sun Jul 14 2013 - version 1.11.1
- Playground VPS: send transactions to remove IPs when VPS is lazy deleted

* Tue Jul 09 2013 - version 1.11.0
- Password changer: fix critical bug, password is randomly generated
- VPS swap: update custom configs only if they differ

* Mon Jul 08 2013 - version 1.10.1
- Offline migration: remove IPs on destination node
- Status page: fix VPS count

* Mon Jul 08 2013 - version 1.10.0
- Own implementation of offline migration
- ZFS support for backups and restore
- VPS clone: fix backup export setting, support new implementation
- VPS swap: adapt to new implementation of offline migration
- NAS mount list: do not show mounts for deleted VPSes
- Playground: unassigned IP addresses from lazy deleted VPS (#39), improve automated server picking
- Restore: fix checking if another restore is already queued

* Mon May 27 2013 - version 1.9.0
- VPS swapping
- Add link to user profile next to logout button
- Show backup sizes
- Add reason field to VPS config change
- VPS mass management: add action restore VPS run state
- Fix backup exclude saving

* Mon May 13 2013 - version 1.8.2
- Mailer: do not alert about expiring VPS that are already deleted

* Mon Apr 15 2013 - version 1.8.1
- VPS clone & reinstall: fixed mounts regeneration
- VPS restore & download: fixed trasaction server id
- Implemented VPS expiration
- Playground: pick node automatically, member can have lazy deleted VPS and create new one,
  set expiration
- Members don't see lazy deleted VPS
- Fixed database upgrading
- Admins can always immediately delete VPS

* Sun Apr 07 2013 - version 1.8.0
- Implement NAS - exports on storage nodes and their mounts on vz nodes
- Backups are built on NAS as well
- Database versioning with automated upgrade
- Lazy deletion of members and VPS
- Reorganized page cluster
- Sysconfig uses JSON instead of (un)serialize
- VPS reinstall: skip IP del
- Traffic accounting: fix IPv6 data, fix table columns
- Transactions: show both time since addition to queue and real execution time

* Tue Feb 12 2013 - version 1.7.3
- Fixed sample config saving
- Clone: fixed not settings features enabled flag
- Set backuper: remove the need of page refresh to see the changes
- Transactions: yet another fix of t_params is always hash, never array

* Sun Feb 10 2013 - version 1.7.2
- Transactions: t_params is always hash, never array

* Sat Feb 09 2013 - version 1.7.1
- Fix optional backup mount in mass management and VPS details
- Configs: add dependency to reapply configs to all affected VPSes on create config
- Clone: create custom config if any

vpsadminctl.rb
#!/usr/bin/env ruby

require 'pathname'

$: << File.dirname(Pathname.new(__FILE__).realpath) unless $:.include? File.dirname(__FILE__)

require 'optparse'
require 'lib/rc'

options = {
	:parsable => false,
	:sock => '/var/run/vpsadmind.sock',
}

command = nil

opt_parser = OptionParser.new do |opts|
	opts.banner = <<END_BANNER
Usage: vpsadminctl <command> [global options] [command options]

Commands:
END_BANNER

  Command.all do |c|
    opts.banner += sprintf("%-20s %s\n", c.label, c.description)
  end

  opts.banner += <<END_BANNER

For specific options type: vpsadminctl <command> --help

END_BANNER

  unless ARGV[0].nil?
    klass = Command.get(ARGV[0])

    if klass
      command = klass.new
      opts.separator 'Specific options:'

      command.options(opts, ARGV[1..-1])
    end
  end
	
	opts.separator ''
	opts.separator 'Global options:'
	
	opts.on('-p', '--parsable', 'Use in scripts, output can be easily parsed') do
		options[:parsable] = true
	end
	
	opts.on('-s', '--socket SOCKET', 'Specify socket') do |s|
		options[:sock] = s
	end
	
	opts.on('-v', '--version', 'Print version and exit') do
		puts VpsAdminCtl::VERSION
		exit
	end
	
	opts.on_tail('-h', '--help', 'Show this message') do
		puts opts
		exit
	end
end

unless ARGV.size > 0
	$stderr.puts 'No command specified'
	puts opt_parser
	exit(false)
end

begin
	opt_parser.parse!
rescue OptionParser::InvalidOption
	$stderr.puts $!
	puts opt_parser
	exit(false)
end

rc = VpsAdminCtl::RemoteControl.new(options)

unless command
	$stderr.puts 'Invalid command'
	puts opt_parser
	exit(false)
end

ret = rc.exec(command)

if ret && ret[:status] == :failed
	$stderr.puts "Command error: #{ret[:error]}"
	exit(false)
end
